File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/longest_valid_parentheses/LongestValidParentheses3.java
package main.java.com.jiang.leetcode100.longest_valid_parentheses;

import java.util.Scanner;
import java.util.Stack;

class Solution3 {
    public int longestValidParentheses(String s) {
        int n = s.length();
        int max_len = 0;
        int left = 0, right = 0;
        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) == '(') left++;
            else if(s.charAt(i) == ')') right++;
            if(left == right) max_len = Math.max(max_len, left * 2);
            else if(left < right) {
                left = 0;
                right = 0;
            }
        }

        left = 0;
        right = 0;
        for(int i = s.length() - 1; i >= 0; i--) {
            if(s.charAt(i) == '(') left++;
            else if(s.charAt(i) == ')') right++;
            if(left == right) max_len = Math.max(max_len, left * 2);
            else if(right < left) {
                left = 0;
                right = 0;
            }
        }

        return max_len;
    }
}

public class LongestValidParentheses3 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();
        scanner.close();
        System.out.println(new Solution3().longestValidParentheses(s));
    }
}

File: Leetcode100/leetcode100_python/LongestValidParentheses3.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        n = len(s)
        max_len = 0
        left, right = 0, 0
        for i in range(n):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                max_len = max(max_len, 2 * left)
            elif left < right:
                left = 0
                right = 0

        left, right = 0, 0
        for i in range(n - 1, -1, -1):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                max_len = max(max_len, 2 * left)
            elif right < left:
                left = 0
                right = 0
        return max_len


if __name__ == "__main__":
    s = input()
    solution = Solution()
    print(solution.longestValidParentheses(s))